### PR TITLE
Cache TCB bundles and lower refreshing frequency

### PR DESCRIPTION
**Pull request by jberci**

### CODE REVIEW

File: go/runtime/host/sgx/ecdsa.go [hunk 1]
```diff
@@ -18,6 +20,19 @@ import (
 	"github.com/oasisprotocol/oasis-core/go/runtime/host/protocol"
 )
 
+const (
+	tcbCacheKey = "tcb bundle cache"
```
Comment (kostko) [inline go/runtime/host/sgx/ecdsa.go:24]:
```suggestion
	tcbCacheKey = "tcb_bundle_cache"
```

File: go/runtime/host/sgx/ecdsa.go [hunk 2]
```diff
@@ -105,6 +110,75 @@ func (ec *teeStateECDSA) Update(ctx context.Context, sp *sgxProvisioner, conn pr
...
+	case nil, persistent.ErrNotFound:
+		// Not cached yet. Not an error.
+	default:
+		// Can't get it... an error, but we can still try downloading it.
+		sp.logger.Warn("error checking common store for cached TCB bundle",
+			"err", err,
+		)
+	}
+
+	if cachedBundle == nil {
```
Comment (kostko) [inline go/runtime/host/sgx/ecdsa.go:142]: It would be good if this bundle-cache logic was in a separate struct so it could be more easily unit-tested with various bundles/timestamps/etc.

File: go/runtime/registry/config.go [hunk 3]
```diff
@@ -188,7 +189,7 @@ func newConfig(dataDir string, consensus consensus.Backend, ias ias.Endpoint) (*
 					Consensus:         consensus,
 					SandboxBinaryPath: sandboxBinary,
 					InsecureNoSandbox: insecureNoSandbox,
-				})
+				}, commonStore)
```
Comment (kostko) [inline go/runtime/registry/config.go:192]: I would rather put this into the `hostSgx.Config` structure, similar to the others.

File: go/runtime/host/sgx/ecdsa.go [hunk 4]
```diff
@@ -66,17 +67,7 @@ func (ec *teeStateECDSA) Init(ctx context.Context, sp *sgxProvisioner, runtimeID
...
-	if err != nil {
-		return nil, fmt.Errorf("failed to get quote: %w", err)
-	}
-
-	var quote pcs.Quote
-	if err = quote.UnmarshalBinary(rawQuote); err != nil {
-		return nil, fmt.Errorf("failed to parse quote: %w", err)
-	}
-
+func (ec *teeStateECDSA) downloadBundle(ctx context.Context, sp *sgxProvisioner, quote pcs.Quote) (*pcs.TCBBundle, error) {
```
Comment (kostko) [inline go/runtime/host/sgx/ecdsa.go:70]: Pass `*pcs.Quote` instead.

File: go/runtime/host/sgx/sgx.go [hunk 5]
```diff
@@ -85,7 +89,7 @@ type teeStateImpl interface {
 	Init(ctx context.Context, sp *sgxProvisioner, runtimeID common.Namespace, version version.Version) ([]byte, error)
 
 	// Update updates the TEE state and returns a new attestation.
-	Update(ctx context.Context, sp *sgxProvisioner, conn protocol.Connection, report []byte, nonce string) ([]byte, error)
+	Update(ctx context.Context, serviceStore *persistent.ServiceStore, sp *sgxProvisioner, conn protocol.Connection, report []byte, nonce string) ([]byte, error)
```
Comment (kostko) [inline go/runtime/host/sgx/sgx.go:92]: Why is this needed given that it is already available in `sp.serviceStore`?

File: go/runtime/host/sgx/tcb_cache.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	tcbCacheSlowRefreshInterval = 24 * time.Hour
+)
+
+type tcbBundleCache struct {
+	bundle         *pcs.TCBBundle `json:"bundle"`
+	expectedExpiry string         `json:"expected_expiry"`
+	lastUpdate     string         `json:"last_update"`
+}
+
+func checkTCBCache(store *persistent.ServiceStore, sp *sgxProvisioner) (bundle *pcs.TCBBundle, refresh bool) {
```
Comment (kostko) [inline go/runtime/host/sgx/tcb_cache.go:24]: Similar you can grab the store from `sp`.

File: go/runtime/host/sgx/sgx.go [hunk 7]
```diff
@@ -35,6 +36,9 @@ const (
 	sandboxMountRuntime   = "/runtime"
 	sandboxMountSignature = "/runtime.sig"
 
+	// The service name for the common store to use for TCB cache bookkeeping.
+	serviceStoreName = "sgx_runtime_host"
```
Comment (kostko) [inline go/runtime/host/sgx/sgx.go:40]:
```suggestion
	serviceStoreName = "runtime_host_sgx"
```

File: go/runtime/host/sgx/tcb_cache.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+			"err", err,
+		)
+	}
+
+	// Decide if the cache needs to be refreshed.
+	refresh = true
+	if stored != nil {
+		refresh = func() bool {
+			var expectedExpiry time.Time
+			if expectedExpiry, err = time.Parse(pcs.TimestampFormat, stored.expectedExpiry); err != nil {
```
Comment (kostko) [inline go/runtime/host/sgx/tcb_cache.go:44]: Why parse these each time instead of already storing it as `time.Time`?

**Review by kostko — COMMENTED** [review]
Nice! Some minor comments.

File: go/runtime/host/sgx/tcb_cache.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	tcbCacheKey = "tcb_bundle_cache"
+
+	tcbCacheRefreshThreshold    = 14 * 24 * time.Hour
+	tcbCacheSlowRefreshInterval = 24 * time.Hour
+)
+
+type tcbBundleCache struct {
+	bundle         *pcs.TCBBundle `json:"bundle"`
+	expectedExpiry time.Time      `json:"expected_expiry"`
+	lastUpdate     time.Time      `json:"last_update"`
```
Comment (kostko) [inline go/runtime/host/sgx/tcb_cache.go:21]: I believe this will not work (e.g. unmarshalling will just zero-initialize these) as private fields cannot be deserialized properly (so you need to capitalize them to export).

Would be good to have some basic tests of the cache that makes sure storage/retrieval works. To make testing easier you could create a structure (e.g. `tcbCache`) that one can create and holds a service store and logger. Then in tests you can more easily (e.g. using a test service store) test it without the need to create an `sgxProvisioner`.

File: go/runtime/host/sgx/ecdsa.go [hunk 10]
```diff
@@ -129,19 +164,33 @@ func (ec *teeStateECDSA) Update(ctx context.Context, sp *sgxProvisioner, conn pr
...
+		tcbBundle = cachedTCBBundle
+	} else {
+		// If downloaded already, don't try again but just return the last error.
+		if freshTCBBundle != nil {
+			return nil, err
+		}
+
+		// If not downloaded yet this time round, try forcing. Any errors are fatal.
+		var downloadedTCBBundle *pcs.TCBBundle
+		if downloadedTCBBundle, err = ec.downloadBundle(ctx, sp, &quote); err != nil {
```
Comment (peternose) [inline go/runtime/host/sgx/ecdsa.go:183]: Feels like we are repeating ourselves. Maybe we merge this part with the upper one.
```
	tcbBundle, err = func() (*pcs.TCBBundle, error) {
		if cachedTCBBundle, err := tcbCache.check(); err == nil {
			if err = ec.verifyBundle(quote, quotePolicy, freshTCBBundle, sp, "cached"); err == nil {
				return cachedTCBBundle, nil
			}
		}

		downloadedTCBBundle, err := ec.downloadBundle(ctx, sp, &quote);
		if err != nil {
			return nil, err
		}
		if err = ec.verifyBundle(quote, quotePolicy, freshTCBBundle, sp, "cached"); err != nil {
			return nil, err
		}
		tcbCache.cache(downloadedTCBBundle)
		return downloadedTCBBundle, nil
	}()
	if err != nil {
		return nil, err
	}
```

**Review by peternose — COMMENTED** [review]
Left few comments, ignore those which you find irrelevant.

File: go/runtime/host/sgx/tcb_cache.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+		}
+	}
+
+	// Decide if the cache needs to be refreshed.
+	refresh = true
+	if stored != nil {
+		refresh = func() bool {
+			now := tc.now()
+
+			// Wait for the first two weeks, then check once daily.
```
[the remark this answers is not in the document]
Comment (jberci) [reply go/runtime/host/sgx/tcb_cache.go]: I think it is: if the bundle is expected to be valid for around a month, then there shouldn't be any fetches for the first two weeks of that, in the last two weeks they should be daily, and after expiration there shouldn't be any throttling.

File: go/runtime/host/sgx/ecdsa.go (hunk 10 above)
Comment (jberci) [reply go/runtime/host/sgx/ecdsa.go]: I'm not sure the snippets are equivalent: the download must not be unconditional, but if it succeeds, its bundle should take precedence over whatever was in the cache.

Comment (peternose) [reply go/runtime/host/sgx/ecdsa.go]: You are right. I thought that refresh is mandatory, but is not, as we use the cached bundle if download fails, which fine.

File: go/runtime/host/sgx/ecdsa.go [hunk 12]
```diff
@@ -129,19 +150,48 @@ func (ec *teeStateECDSA) Update(ctx context.Context, sp *sgxProvisioner, conn pr
...
+				sp.logger.Warn("error downloading TCB refresh",
+					"err", err,
+				)
+			}
+		}
+
+		if err = ec.verifyBundle(quote, quotePolicy, fresh, sp, "fresh"); err == nil {
+			ec.tcbCache.cache(fresh)
+			return fresh, nil
+		} else if err = ec.verifyBundle(quote, quotePolicy, cached, sp, "cached"); err == nil {
```
Comment (peternose) [inline go/runtime/host/sgx/ecdsa.go:171]:
```suggestion
		}
		if err = ec.verifyBundle(quote, quotePolicy, cached, sp, "cached"); err == nil {
```

File: go/runtime/host/sgx/ecdsa.go [hunk 13]
```diff
@@ -129,19 +150,48 @@ func (ec *teeStateECDSA) Update(ctx context.Context, sp *sgxProvisioner, conn pr
...
+			return cached, nil
+		} else {
```
Comment (peternose) [inline go/runtime/host/sgx/ecdsa.go:173]:
```suggestion
```

**Review by peternose — APPROVED** [review]

File: go/runtime/host/sgx/tcb_cache.go [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+	var err error
+
+	// Check if we have a copy in the local store.
+	var stored *tcbBundleCache
+	switch err = tc.serviceStore.GetCBOR([]byte(tcbCacheKey), &stored); err {
+	case nil, persistent.ErrNotFound:
+		// Not cached yet. Not an error.
+	default:
+		// Can't get it... an error, but we can still try downloading it.
+		if tc.logger != nil {
```
[the remark this answers is not in the document]
Comment (ptrus) [reply go/runtime/host/sgx/tcb_cache.go:41]: Or also create the logger in the `mockTbcCache`. I think logging an error here makes sense, since this is an internal to cache error, that's intended for the node operator not for the calling code.

File: go/runtime/host/sgx/tcb_cache.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+			tc.logger.Warn("error checking common store for cached TCB bundle",
+				"err", err,
+			)
+		}
+	}
+
+	// Decide if the cache needs to be refreshed.
+	if err != nil {
```
Comment (ptrus) [inline go/runtime/host/sgx/tcb_cache.go:49]: I think it would be clearer if this conditional is moved into the above switch? e.g.
```golang
switch err {
  case nil:
     // No error, continues below.
  case persistent.ErrNotFound:
    // Not cached yet, not an error, but needs refresh.
    return nil, true
  default:
    // Log the error and needs refresh.
    ...
    return nil, true
}
```

**Review by peternose — APPROVED** [review]
Now, we can probably remove this [line](https://github.com/oasisprotocol/oasis-core/blob/master/go/common/sgx/pcs/http.go#L88). 

Tested on an sgx instance. The bundle is downloaded, cached and used on restarts :+1:

**Review by kostko — APPROVED** [review]

**Review by kostko — APPROVED** [review]
